token/token.go

```diff
@@ -1,75 +1,146 @@
+	GivenName     string                `json:"given_name"`
+	FamilyName    string                `json:"family_name"`
+	Email         string                `json:"email"`
+	Company       string                `json:"company"`
+	SessionState  string                `json:"session_state"`
+	Authorization *AuthorizationPayload `json:"authorization"`
+	jwt.StandardClaims
+}
+
+// AuthorizationPayload represents an authz payload in the rpt token
+type AuthorizationPayload struct {
+	Permissions []Permissions `json:"permissions"`
+}
+
+// Permissions represents a "permissions" in the AuthorizationPayload
+type Permissions struct {
+	ResourceSetName *string `json:"resource_set_name"`
+	ResourceSetID   *string `json:"resource_set_id"`
+}
+
+type PublicKey struct {
+	KeyID string
+	Key   *rsa.PublicKey
+}
+
 // Manager generate and find auth token information
 type Manager interface {
-	Extract(string) (*account.Identity, error)
 	Locate(ctx context.Context) (uuid.UUID, error)
-	PublicKey() *rsa.PublicKey
+	ParseToken(ctx context.Context, tokenString string) (*TokenClaims, error)
+	PublicKey(kid string) *rsa.PublicKey
+	PublicKeys() []*rsa.PublicKey
 }
 
 type tokenManager struct {
-	publicKey  *rsa.PublicKey
-	privateKey *rsa.PrivateKey
+	publicKeysMap map[string]*rsa.PublicKey
+	publicKeys    []*PublicKey
 }
 
 // NewManager returns a new token Manager for handling tokens
-func NewManager(publicKey *rsa.PublicKey) Manager {
-	return &tokenManager{
-		publicKey: publicKey,
+func NewManager(config configuration) (Manager, error) {
+	// Load public keys from Auth service and add them to the manager
+	tm := &tokenManager{
+		publicKeysMap: map[string]*rsa.PublicKey{},
 	}
-}
 
-// NewManagerWithPrivateKey returns a new token Manager for handling creation of tokens with both private and pulic keys
-func NewManagerWithPrivateKey(privateKey *rsa.PrivateKey) Manager {
-	return &tokenManager{
-		publicKey:  &privateKey.PublicKey,
-		privateKey: privateKey,
-	}
-}
-
-func (mgm tokenManager) Extract(tokenString string) (*account.Identity, error) {
-	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
-		return mgm.publicKey, nil
-	})
+	remoteKeys, err := authtoken.FetchKeys(config.GetKeysEndpoint())
 	if err != nil {
-		return nil, errors.WithStack(err)
+		log.Error(nil, map[string]interface{}{
+			"keys_url": config.GetKeysEndpoint(),
+		}, "unable to load public keys from remote service")
+		return nil, errors.New("unable to load public keys from remote service")
+	}
+	for _, remoteKey := range remoteKeys {
+		tm.publicKeysMap[remoteKey.KeyID] = remoteKey.Key
+		tm.publicKeys = append(tm.publicKeys, &PublicKey{KeyID: remoteKey.KeyID, Key: remoteKey.Key})
+		log.Info(nil, map[string]interface{}{
+			"kid": remoteKey.KeyID,
+		}, "Public key added")
 	}
 
-	if !token.Valid {
-		return nil, errors.New("Token not valid")
+	devModeURL := config.GetKeycloakDevModeURL()
+	if devModeURL != "" {
```

Comment:
aren't you doing the same work as above, with another KC URL ?

Reply:
Not exactly. Main difference is that here we are operating with different PublicKey type. (fabric8-auth/PublicKey vs. fabric8-wit/PublicKey). We could re-use this code in both places but it would require to introduce some common interface which can be tricky. So, anyway, it would require us to write more code to re-use this small block in two places. Which doesn't make sense.